File: backend/app/routers/messages.py
from typing import List, Optional
from datetime import datetime
from fastapi import APIRouter, HTTPException, status, Depends, Query
from bson import ObjectId
from app.database import get_database
from app.models.message import MessageCreate, Message
from app.utils.auth import get_current_user_id
from app.utils.trip_permissions import check_trip_access





router = APIRouter(prefix="/trips/{trip_id}/messages", tags=["messages"])




@router.get("", response_model=List[dict])
async def get_messages(
    trip_id: str,
    limit: int = Query(default=200, le=500),
    cursor: Optional[str] = None,
    user_id: str = Depends(get_current_user_id)
):
    """Get messages for a trip"""
    await check_trip_access(trip_id, user_id)
    
    db = get_database()
    query = {"tripId": ObjectId(trip_id)}
    
    if cursor:
        try:
            cursor_time = ObjectId(cursor)
            query["_id"] = {"$gt": cursor_time}
        except:
            pass
    
    messages = await db.messages.find(query).sort("createdAt", 1).limit(limit).to_list(limit)
    
    return [
        {
            "id": str(msg["_id"]),
            "trip_id": str(msg["tripId"]),
            "author_id": str(msg["authorId"]) if msg.get("authorId") else None,
            "type": msg["type"],
            "content": msg["content"],
            "summary": msg.get("summary"),
            "questions": msg.get("questions"),
            "conflict_id": str(msg["conflictId"]) if msg.get("conflictId") else None,
            "has_view_plan": msg.get("hasViewPlan", False),
            "created_at": msg.get("createdAt"),
        }
        for msg in messages
    ]


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_message(
    trip_id: str,
    message_data: MessageCreate,
    user_id: str = Depends(get_current_user_id)
):
    """Create a new message"""
    await check_trip_access(trip_id, user_id)
    
    db = get_database()
    
    now = datetime.utcnow()
    message_doc = {
        "tripId": ObjectId(trip_id),
        "authorId": ObjectId(user_id),
        "type": message_data.type,
        "content": message_data.content,
        "summary": message_data.summary,
        "questions": message_data.questions,
        "hasViewPlan": message_data.has_view_plan,
        "createdAt": now,
    }
    
    result = await db.messages.insert_one(message_doc)
    message_doc["_id"] = result.inserted_id
    
    # Update trip's updated_at
    await db.trips.update_one(
        {"_id": ObjectId(trip_id)},
        {"$set": {"updatedAt": now}}
    )
    
    return {
        "id": str(message_doc["_id"]),
        "trip_id": trip_id,
        "author_id": user_id,
        "type": message_doc["type"],
        "content": message_doc["content"],
        "summary": message_doc.get("summary"),
        "questions": message_doc.get("questions"),
        "has_view_plan": message_doc.get("hasViewPlan", False),
        "created_at": now,
    }

File: backend/src/routers/messages.ts
import { Router, Response } from 'express';
import { ObjectId } from 'mongodb';
import { getDatabase } from '../database.js';
import { MessageCreateSchema } from '../models/message.js';
import { getCurrentUserId, AuthRequest } from '../utils/auth.js';
import { checkTripAccess } from '../utils/trip_permissions.js';

const router = Router({ mergeParams: true });

router.get('', getCurrentUserId, async (req: AuthRequest, res: Response) => {
  try {
    const { trip_id } = req.params;
    const limit = Math.min(parseInt(req.query.limit as string || '200', 10), 500);
    const cursor = req.query.cursor as string | undefined;
    
    await checkTripAccess(trip_id, req.userId!);
    
    const db = getDatabase();
    const query: any = { tripId: new ObjectId(trip_id) };
    
    if (cursor) {
      try {
        query._id = { $gt: new ObjectId(cursor) };
      } catch {
        // Invalid cursor, ignore
      }
    }
    
    const messages = await db.collection('messages')
      .find(query)
      .sort({ createdAt: 1 })
      .limit(limit)
      .toArray();
    
    res.json(messages.map(msg => ({
      id: msg._id.toString(),
      trip_id: msg.tripId.toString(),
      author_id: msg.authorId?.toString() || null,
      type: msg.type,
      content: msg.content,
      summary: msg.summary,
      questions: msg.questions,
      conflict_id: msg.conflictId?.toString() || null,
      plan_version_id: msg.planVersionId?.toString() || null,
      has_view_plan: msg.hasViewPlan || false,
      created_at: msg.createdAt,
    })));
  } catch (error: any) {
    if (error.message === 'Trip not found' || error.message === 'Access denied') {
      res.status(error.message === 'Trip not found' ? 404 : 403).json({ detail: error.message });
    } else {
      res.status(500).json({ detail: error.message });
    }
  }
});

router.post('', getCurrentUserId, async (req: AuthRequest, res: Response) => {
  try {
    const { trip_id } = req.params;
    const validated = MessageCreateSchema.parse(req.body);
    
    await checkTripAccess(trip_id, req.userId!);
    
    const db = getDatabase();
    const now = new Date();
    const messageDoc = {
      tripId: new ObjectId(trip_id),
      authorId: validated.type === 'human' ? new ObjectId(req.userId!) : null,
      type: validated.type,
      content: validated.content,
      summary: validated.summary,
      questions: validated.questions,
      hasViewPlan: validated.has_view_plan,
      planVersionId: validated.plan_version_id ? new ObjectId(validated.plan_version_id) : undefined,
      createdAt: now,
    };
    
    const result = await db.collection('messages').insertOne(messageDoc);
    
    // Update trip's updated_at
    await db.collection('trips').updateOne(
      { _id: new ObjectId(trip_id) },
      { $set: { updatedAt: now } }
    );
    
    res.status(201).json({
      id: result.insertedId.toString(),
      trip_id,
      author_id: messageDoc.authorId?.toString() || null,
      type: messageDoc.type,
      content: messageDoc.content,
      summary: messageDoc.summary,
      questions: messageDoc.questions,
      conflict_id: null,
      plan_version_id: messageDoc.planVersionId?.toString() || null,
      has_view_plan: messageDoc.hasViewPlan,
      created_at: now,
    });
  } catch (error: any) {
    if (error.name === 'ZodError') {
      res.status(400).json({ detail: 'Validation error', errors: error.errors });
      return;
    }
    if (error.message === 'Trip not found' || error.message === 'Access denied') {
      res.status(error.message === 'Trip not found' ? 404 : 403).json({ detail: error.message });
    } else {
      res.status(500).json({ detail: error.message });
    }
  }
});

export default router;
